Bury bip9 deployments

jnewbery opened the pull request:
This hardcodes CSV and segwit activation heights, similar to the BIP 90 buried deployments for BIPs 34, 65 and 66.

CSV and segwit have been active for over 18 months. Hardcoding the activation height is a code simplification, makes it easier to understand segwit activation status, and reduces technical debt.

This was originally attempted by jl2012 in #11398 and again by me in #12360.

jamesob committed:
[tests] Add coverage for the content of getblockchaininfo.softforks

jnewbery committed:
[rpc] Tidy up reporting of buried and ongoing softforks

This combines reporting of buried (formally ISM) softfork deployments
and BIP9 versionbits softfork deployments into one JSON object in the
getblockchaininfo return object.

jnewbery committed:
[Consensus] Bury CSV deployment height

Hard code CSV deployment height to 419328 for mainnet.

jnewbery committed:
[Consensus] Bury segwit deployment

Hardcode segwit deployment height to 481824 for mainnet.

jnewbery commented:
@jl2012 originally attempted this in #11398, which had concept ACKs from @gmaxwell @dcousens and @sdaftuar . I then tried again in #12360, which got ACKs from @jamesob and @jtimon .

practicalswift commented:
Concept ACK

sipa commented:
Concept ACK

MarcoFalke commented:
Wouldn't it help to split out the first commit, which reports ISM deployments in a different way? It is a breaking rpc change and probably needs different review than a consensus refactor (as well as release notes)

jnewbery commented:
Force-pushed to fix some minor issues from rebasing.

> Wouldn't it help to split out the first commit, which changes how ISM deployments are reported?

All three commits change the output of `getblockchaininfo` and how soft forks are reported, so it makes sense to merge them in the same PR.

laanwj commented:
Concept ACK

PastaPastaPasta commented on src/chainparams.cpp:
```diff
@@ -359,11 +339,23 @@ class CRegTestParams : public CChainParams {
+        if (height < -1 || height >= std::numeric_limits<int>::max()) {
+            throw std::runtime_error(strprintf("Activation height %ld for segwit is out of valid range. Use -1 to disable segwit.", height));
+        }
+        else if (height == -1) {
```
based on https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c it appears this line should be on the line above with the bracket.
`} else if (height == -1) {`

jnewbery replied:
fixed

PastaPastaPasta commented on src/rpc/blockchain.cpp:
```diff
@@ -1278,18 +1270,21 @@ static UniValue BIP9SoftForkDesc(const Consensus::Params& consensusParams, Conse
+    if (ThresholdState::LOCKED_IN == thresholdState) {
+        rv.pushKV("height", since_height + consensusParams.nMinerConfirmationWindow);
+    }
+    else if (ThresholdState::ACTIVE == thresholdState) {
```
same as above 
based on https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c it appears this line should be on the line above with the bracket.

jnewbery replied:
fixed

jnewbery commented:
Thanks for the review @PastaPastaPasta . I've fixed your two comments.

laanwj commented on src/chainparams.cpp:
```diff
@@ -69,6 +69,8 @@ class CMainParams : public CChainParams {
         consensus.BIP65Height = 388381; // 000000000000000004c2b624ed5d7756c508d90fd0da2c7c679febfa6c4735f0
         consensus.BIP66Height = 363725; // 00000000000000000379eaa19dce8c9b722d46ae6a57c2f1a988119488b50931
+        consensus.CSVHeight = 419328; // 000000000000000004a1b34462cb8aeebd5799177f7a29cf28f2d1961716b5b5
+        consensus.SegwitHeight = 481824; // 0000000000000000001c8018d9cb3b742ef25114f27563e3fc4a1902167f9893
```
Checked blockhashes against bock chain, and checked the activation heights against `getblockchaininfo` that they match `since`
```json
  "bip9_softforks": {
    "csv": {
      "status": "active",
      "startTime": 1462060800,
      "timeout": 1493596800,
      "since": 419328
    },
    "segwit": {
      "status": "active",
      "startTime": 1479168000,
      "timeout": 1510704000,
      "since": 481824
    }
  },
```

laanwj replied:
It's slightly inconsistent how all of these (BIP34, BIP65, BIP66) are BIPxx and the new ones are written out as name. I don't personally mind, to be honest, I even think this is clearer than having numbers.

jnewbery replied:
Neither the CSV nor segwit softforks were specified by a single BIP. CSV was BIPs 68, 112, and 113 and segwit was BIPs 141, 143, and 147. We could go back and give BIP34, BIP65 and BIP66 'descriptive' names, but I think that's an unnecessary breaking change.

laanwj commented on src/rpc/blockchain.cpp:
```diff
@@ -1315,29 +1309,25 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
-            "     }, ...\n"
-            "  ],\n"
-            "  \"bip9_softforks\": {           (object) status of BIP9 softforks in progress\n"
+            "  \"softforks\": {                (object) status of softforks\n"
```
I like making this a object indexed by name, makes it easier to look one up in languages that treat them as dictionaries.

promag replied:
Agree, but this is a breaking change, seems unnecessary.

jnewbery replied:
> I like making this a object indexed by name, makes it easier to look one up in languages that treat them as dictionaries.

For the benefit of other reviewers - that's what this PR does already.

> this is a breaking change, seems unnecessary

I doubt that there are many clients programmatically querying this, since the value never changes once it's locked in. After this PR, there don't need to be further breaking changes, since the new `softfork` object is able to hold both bip9 and buried fork details (or any other potential activation method types).

promag replied:
We are on the same page then, but for the sake of complaints, just add a small release note? @MarcoFalke also asked it above.

merehap commented on src/validation.cpp:
```diff
@@ -3157,14 +3157,14 @@ bool CheckBlock(const CBlock& block, CValidationState& state, const Consensus::P
 
 bool IsNullDummyEnabled(const CBlockIndex* pindexPrev, const Consensus::Params& params)
 {
     LOCK(cs_main);
```
Can this LOCK(cs_main) be removed just like the one in IsWitnessEnabled was? If it shouldn't be, the reason would be unintuitive, so could you leave a comment why it is still needed/desirable in that case?

jnewbery replied:
Yes, it can. In fact, this function is only called from within `GetBlockScriptFlags()`, which holds `cs_main`. After this PR, this is now a simple one-line check, so I've removed the function and placed the conditional directly into `GetBlockScriptFlags()`.

jtimon commented on src/chainparams.cpp:
```diff
@@ -349,11 +339,22 @@ class CRegTestParams : public CChainParams {
-void CRegTestParams::UpdateVersionBitsParametersFromArgs(const ArgsManager& args)
+void CRegTestParams::UpdateActivationParametersFromArgs(const ArgsManager& args)
 {
+    if (gArgs.IsArgSet("-segwitheight")) {
```
It would be nicer to do something more generic that works for other deployments, not just segwit. For example, if anybody was relying on vparams to test csv activation, it won't be possible after this PR.
But perhaps this is out of the scope for this PR, 
With something like https://github.com/bitcoin/bitcoin/pull/8994 we could simply make the all the buried heights configurable for custom regtests.

jtimon replied:
Perhaps we cna call the parameter "-con_segwitheight: like in #8994 ?

jnewbery replied:
This argument is just for testing, so it's not disruptive to change this in future if we think the `-con_xxx` naming scheme is better. For now, I'd like to keep this as it is, rather than optimizing for a PR that hasn't been merged yet.

jtimon commented:
I didn't check the actual activation heights, but beyond that and my small nit, utACK f395b23aa0d125aa9534ff05df8d7ccd2d576868

meshcollider commented:
Concept ACK

moneyball commented:
I've looked through the history of the 3 PRs trying to get this change merged, and there is widespread support for it, but we need to get it to the finish line! Now that it has been added to the high priority list, would one or two of the following folks be able to review this in the next week or so? @sdaftuar @TheBlueMatt @sipa @jl2012 🙏

instagibbs commented:
just noting there is a competing proposal at #16229

laanwj commented:
> just noting there is a competing proposal at #16229

Why are there competing proposals? Is there a comparison sometimes of the pros/cons of either approach?

jnewbery commented:
This needs release notes if it reaches a stage where it's ready for merge.

ajtowns commented:
> Why are there competing proposals? Is there a comparison sometimes of the pros/cons of either approach?

The propsals take different approaches: this one (#16060) changes segwit/csv from bip9 deployments to be hardcoded fixed height deployments; the other one (#16229) generalises bip9 deployments to also support fixed height deployments, and converts strictder and cltv into this generalised deployment method and updates segwit/csv to be fixed height rather than bip9 based.

I think the benefits of hardcoded fixed height stuff is that it's a bit more efficient (the height's hardcoded and every codepath knows the height is hardcoded), while the cost is that anytime you change a bip9 deployment to a fixed height deployment you have to change the code in a few places. Doing it this way also means you can't have tests that want to control how the forks activated, which results in substantive changes to the test suite when burying a deployment.

The cost of the generalised approach is that you can't assume that the deployments are just activated at a particular height, which slows mainnet handling of the deployments down for (at most) a benefit in regtest. I think this slowdown is pretty trivial, but haven't benchmarked it. The benefits (I think) are that it means we can bury BIP9 deployments in future just by changing the declaration in chainparams.cpp, with tests generally continuing to work unchanged, and no code changes needed for getblockchaininfo or in checking when the soft-fork is enabled. This approach can also be done with minimal changes to getblockchaininfo output if we thought that was a benefit.

jnewbery commented:
> The propsals take different approaches....

Thanks @ajtowns - that seems like a good summary to me. I left a comment in your PR about why I prefer the approach here, which I'll copy below:

---

I prefer the approach in #16060. The changes in this PR seem more complex for achieving a similar goal (+110 lines here vs -70 lines in 16060)

> I think this should make future burials much simpler (it becomes a one-line change)

Softfork activation/burials are sufficiently rare that I don't think we have to optimize to minimize the code changes when they do happen.

> it doesn't change RPC or tests much, and I think it will make it easier to make future changes to activation methods should we wish to do so.

I like the new RPC return object softfork in 16060. It seems simple, non-breaking when a deployment goes from bip9 to buried (the type field determines which other fields or sub-objects appear in the softfork object), and allows for other deployment methods (just add a new type).

We've been trying to bury these deployments for almost two years now (see #11398), so I just want to see it done. If people prefer this approach, I'm happy to close 16060 in favour of this.

jonasschnelli commented:
Concept ACK

jnewbery commented:
Rebased

This PR (or earlier versions) have concept ACKs from:

- @gmaxwell
- @dcousens
- @sdaftuar
- @jamesob
- @jtimon
- @practicalswift
- @sipa
- @laanwj
- @meshcollider
- @jonasschnelli

I'd really appreciate some code review to get this unblocked.

sdaftuar commented on src/rpc/blockchain.cpp:
```diff
@@ -1342,16 +1332,14 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
+    BuriedForkDescPushBack(softforks, "bip65", consensusParams.BIP65Height);
+    BIP9SoftForkDescPushBack(softforks, "csv", consensusParams, Consensus::DEPLOYMENT_CSV);
+    BIP9SoftForkDescPushBack(softforks, "segwit", consensusParams, Consensus::DEPLOYMENT_SEGWIT);
+    BIP9SoftForkDescPushBack(softforks, "testdummy", consensusParams, Consensus::DEPLOYMENT_TESTDUMMY);
```
Shouldn't we just remove this line?  This seems unnecessary...

jnewbery replied:
I think it's ok to leave this for testing BIP9 logic.

sdaftuar commented on src/rpc/blockchain.cpp:
```diff
@@ -1275,29 +1269,25 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
-            "           \"count\": xx,         (numeric) the number of blocks with the version bit set in the current period \n"
-            "           \"possible\": xx       (boolean) returns false if there are not enough blocks left in this period to pass activation threshold \n"
-            "        }\n"
+            "        \"type\": \"xxxx\",         (string) one of \"buried\", \"bip9\"\n"
```
I think this concept of "type" is slightly weird -- we're leaking an implementation detail that shouldn't have any bearing on how this information is used.  It seems strange to me that we'd distinguish between long-since-activated bip9 softforks, and buried softforks, when reporting to users.  Is the goal just to be able to indicate which entries will have a bip9 field and which ones won't?

ajtowns replied:
We leak the details on how we implement the fork deployment anyway -- at worst you could `invalidateblock` back to before the fork activated and query the fork state to see if it's buried or bip9. I think fork activation is important/risky enough that it's worthwhile explicitly revealing how it's being determined.

If we were to do a bip-16-like deployment or a fixed-height-after-bip9-expiry UASF deployment as suggested in the Great Consensus Cleanup bip proposal, I think using the type field to differentiate between old, buried, fixed height deployments and future ones would make sense.

sdaftuar commented on src/rpc/blockchain.cpp:
```diff
@@ -1183,54 +1183,46 @@ static UniValue verifychain(const JSONRPCRequest& request)
+    // Deployments (e.g. testdummy) with timeout value before Jan 1, 2009 are hidden.
+    // A timeout value of 0 guarantees a softfork will never be activated.
+    // This is used when softfork codes are merged without specifying the deployment schedule.
+    if (consensusParams.vDeployments[id].nTimeout <= 1230768000) return;
```
If we get rid of this filtering and the needless invocation of this function on "testdummy" I think things would be simpler.

Actually, it looks like you do this so that you can test the RPC in regtest even after everything else is buried?  I guess code coverage of the RPC is a good reason to do it this way, if that's what you have in mind.

jnewbery replied:
Leaving this in so we can continue to test BIP 9 logic.

sdaftuar commented on src/rpc/blockchain.cpp:
```diff
@@ -1275,29 +1269,25 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
+            "              \"possible\": xx    (boolean) returns false if there are not enough blocks left in this period to pass activation threshold \n"
+            "           }\n"
+            "        },\n"
+            "        \"height\": \"xxxxxx\",     (numeric) height of the first block which the rules are enforced (only for \"buried\" type, or \"bip9\" type with \"locked_in\" or \"active\" status)\n"
```
"are or will be enforced"?  That seems like more accurate phrasing if you're going to do this for LOCKED_IN rules.

jnewbery replied:
Fixed

sdaftuar commented on src/rpc/blockchain.cpp:
```diff
@@ -1275,29 +1269,25 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
+            "           }\n"
+            "        },\n"
+            "        \"height\": \"xxxxxx\",     (numeric) height of the first block which the rules are enforced (only for \"buried\" type, or \"bip9\" type with \"locked_in\" or \"active\" status)\n"
+            "        \"active\": xx,           (boolean) true if the rules are enforced\n"
```
Maybe "true if the rules are enforced in the block that is the current chain tip" to be more explicit about how to think about this for buried deployments?

jnewbery replied:
fixed

sdaftuar commented on src/consensus/params.h:
```diff
@@ -59,6 +58,8 @@ struct Params {
     int BIP66Height;
     /** Block height at which CSV (BIP68, BIP112 and BIP113) becomes active */
     int CSVHeight;
+    /** Block height at which Segwit (BIP141, BIP143 and BIP147) becomes active */
```
I wonder if it's worth adding a comment here to explain that segwit v0 script rules are enforced on all blocks, other than the BIP 16 exception?

sdaftuar commented on src/rpc/blockchain.cpp:
```diff
@@ -1240,18 +1232,20 @@ static UniValue BIP9SoftForkDesc(const Consensus::Params& consensusParams, Conse
+    rv.pushKV("type", "bip9");
+    rv.pushKV("bip9", bip9);
+    if (ThresholdState::LOCKED_IN == thresholdState) {
+        rv.pushKV("height", since_height + consensusParams.nMinerConfirmationWindow);
```
I don't love that we're leaking bip9 implementation calculations into the rpc code.  Though it is useful to see when a LOCKED_IN softfork will actually activate.  Can we invoke a method in the bip9 machinery instead?

jnewbery replied:
Done. Added a `VersionBitsActivationHeight()` function to validation.cpp

sdaftuar reviewed (commented):
Concept ACK, left a bunch of comments but I need to go through the code more carefully still.

I went back and forth on the first commit, because I don't like breaking interfaces, but overall I think it's a big enough improvement that it's worth doing the overhaul this PR proposes (though I did nitpick a few details! which is partly why I hope we don't revisit this RPC output anytime soon after this, as there's infinite room for bikeshedding).

jnewbery reviewed (commented):
I think I've addressed all of @sdaftuar's comments.

maflcko reviewed (commented):
Just some style questions. Feel free to ignore/hide the comments.

I am +-0 on the overall change. So consider my "concept nack" in the previous pull withdrawn, since I don't want to block this pull. I like that it is cleaning up technical debt a bit, but I also think it doesn't improve user experience significantly, which is why I am +-0 on it.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1183,54 +1183,46 @@ static UniValue verifychain(const JSONRPCRequest& request)
-    rv.pushKV("since", VersionBitsTipStateSinceHeight(consensusParams, id));
+    bip9.pushKV("startTime", consensusParams.vDeployments[id].nStartTime);
+    bip9.pushKV("timeout", consensusParams.vDeployments[id].nTimeout);
+    int64_t since_height = VersionBitsTipStateSinceHeight(consensusParams, id);
```
`cs_main` is held for the duration of `getblockchaininfo()`. That's not changed by this PR.

MarcoFalke replied:
Indeed, you are right.

style-nit: You could add `EXCLUSIVE_LOCKS_REQUIRED` to `BIP9SoftForkDescPushBack` and it should still compile.

jnewbery replied:
Done. I've added `EXCLUSIVE_LOCKS_REQUIRED(cs_main)` to `BuriedForkDescPushBack()` and `BIP9SoftForkDescPushBack()`.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1342,16 +1332,14 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
-    softforks.push_back(SoftForkDesc("bip66", 3, tip, consensusParams));
-    softforks.push_back(SoftForkDesc("bip65", 4, tip, consensusParams));
-    for (int pos = Consensus::DEPLOYMENT_CSV; pos != Consensus::MAX_VERSION_BITS_DEPLOYMENTS; ++pos) {
-        BIP9SoftForkDescPushBack(bip9_softforks, consensusParams, static_cast<Consensus::DeploymentPos>(pos));
```
Thanks for digging that out. I really don't understand why that PR was merged. The author's justification (https://github.com/bitcoin/bitcoin/pull/10874#issuecomment-321609838) was "I had only made the change since I needed to add stats about BIP 91 to one of my branches and the fact that the forks needed to be hard coded in to be displayed with getblockchaininfo bothered me." It doesn't make sense to me that lines of code were added to bitcoin core to support someone's fork of the project.

When there are only three soft fork deployments, listing them out explicitly is less lines of code and seems clearer to me.

jnewbery replied:
(also, at the end of this PR, there's only one BIP9 softfork left (TESTDUMMY), so having a for loop _really_ doesn't make sense)

jnewbery commented on src/test/setup_common.cpp:
```diff
@@ -117,9 +117,7 @@ TestingSetup::~TestingSetup()
 TestChain100Setup::TestChain100Setup() : TestingSetup(CBaseChainParams::REGTEST)
 {
-    // CreateAndProcessBlock() does not support building SegWit blocks, so don't activate in these tests.
-    // TODO: fix the code to support SegWit blocks.
```
oops. Bad rebase. I've readded the message.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1275,29 +1269,25 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
-            "     }, ...\n"
-            "  ],\n"
-            "  \"bip9_softforks\": {           (object) status of BIP9 softforks in progress\n"
+            "  \"softforks\": {                (object) status of softforks\n"
```
I think some users _may_ be interested in which height the node started enforcing rules. Even though that's leaking some implementation detail (see https://github.com/bitcoin/bitcoin/pull/16060#discussion_r306528190), I think it's better to show more to the user than less.

As for adding a new RPC, that's even more of a breaking change for clients. For me, it makes most sense to return all blockchain info to the client in a single object from `getblockchaininfo`.

As Suhas points out (https://github.com/bitcoin/bitcoin/pull/16060#pullrequestreview-264351017), there's infinite room for bikeshedding here. Unless there's a really compelling reason to add a separate RPC for softfork deployments, I'd prefer to keep this as it is.

jnewbery commented:
Thanks for the review @MarcoFalke . I believe I've addressed all your comments.

jnewbery commented:
Oops. I dropped a change to the help text (here: https://github.com/bitcoin/bitcoin/pull/16060#discussion_r305496691). Force pushed to get those changes.

ajtowns commented on src/rpc/mining.cpp:
```diff
@@ -482,9 +482,8 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
-    const struct VBDeploymentInfo& segwit_info = VersionBitsDeploymentInfo[Consensus::DEPLOYMENT_SEGWIT];
     // GBT must be called with 'segwit' set in the rules
-    if (setClientRules.count(segwit_info.name) != 1) {
+    if (setClientRules.count("segwit") != 1) {
```
I don't think this is sufficient -- getblocktemplate only populates `aRules` based on versionbits state so whether or not "segwit" is passed as an arg, it ends up outputting an empty array `{"rules": []}` for the template. I think that conflicts with [BIP145](https://github.com/bitcoin/bips/blob/master/bip-0145.mediawiki): "Transactions with witness data may only be included if the template's "rules" list (see BIP 9) includes "segwit"."

jnewbery replied:
I think we're not compliant with this part of BIP 145 anyway, specifically:

> The '!' rule prefix MUST be enabled on the "segwit" rule for templates including transactions with witness data.

There's no code to insert the '!' in templates with blocks containing segwit transactions. I've also tested on master that `getblocktemplate` doesn't include `!segwit` as a rule for templates with witness data.

The fact that this has always been broken suggests to me that no-one is looking at the `rules` field on the return object.

I can add some code to update the `rules` based on whether csv and segwit are activated, but it doesn't seem worth it.

ajtowns replied:
I don't think it makes much sense to require `segwit` in the params if we're not going to bother reporting it ourselves. I think the additional code would just be:

```
    if (pindex->nHeight+1 >= consensusparams.CSVHeight) aRules.push_back("csv");
    if (!fPreSegWit) aRules.push_back("segwit");
```

(also, could have said `fPreSegWit = !IsWitnessEnabled(pindexPrev, consensusParams);`)

Looks like including the `!` got broken prior to activation in March 2017 in #9955 with segwit's `gbt_force` value changing from false to true, so bitcoind hasn't been compliant with bip145 since 0.15, though 0.14 should have been.

I think the logic here isn't right and we should have two settings in `VBDeploymentInfo`, `gbt_simplified` which just adds the `!` when false, and `gbt_optional` which, when the client doesn't specify the rule, avoids setting the versionbit while signalling if `gbt_optional` is true, and errors out when the rule is activated and `gbt_optional` is false. segwit should initially have had `gbt_simplified=false` and `gbt_optional=true`, and just changed `gbt_optional=false` in #14811.

There's no provision in bip9 or bip145 for ever removing forks from the rules list, which might be worth improving, as an alternative.

@luke-jr -- thoughts? You added the gbt changes for bip9 and acked 9955 :)

jnewbery replied:
I can't see how #9955 changed including the `!`. Am I missing something?

jnewbery replied:
ah, I see: https://github.com/bitcoin/bitcoin/pull/9955/files#diff-d6954440f0346f8e42dcd669dd4aeafaL20

'force' is defined as "Whether GBT clients can safely ignore this rule in simplified usage" which seems backwards to me.

luke-jr replied:
I guess `gbt_force` is too unclear a parameter name...

We should fix this. Maybe in a separate PR.

ajtowns reviewed (commented):
I think I'm about -0.3 on this PR (say -0.8 on the approach, +0.5 on the implementation given the approach). So not an ACK, but not a NACK either. (Well, I think there's a bug in the getblocktemplate changes as ~~above~~ below, but presumably that'll be fixed if it is a bug)

The rationale for it is essentially a refactoring: "Hardcoding the activation height is a code simplification, makes it easier to understand segwit activation status, and reduces technical debt." but most of this PR isn't just refactoring:

 * It includes substantial user-visible changes, "breaking" the getblockchaininfo API
 * It drops chunks of functional test coverage, in particular, after this PR, I don't think we have functional tests checking BIP9 activation works anymore

It moves from a somewhat generic interface for soft-fork changes to per-fork code, making it harder to change the way soft-forks are checked in future, which seems like it increased technical debt to me -- doing a generic interface for reporting soft-fork activation will require reverting all the explicit Height member variables and checks.

I think it also turns out to be a more invasive change than necessary, as evidenced by the number of rebases this PR has required, compared to the lack of rebases the alternative PR #16229 has needed. Honestly, I find that a bit surprising, unless the rebasing has been mostly due to the test case changes somehow?

If we want to change the way getblockchaininfo reports softforks (and I think we do), we should split that out as a separate PR (like #16328 proposed) and review those changes independently, rather than forcing them in as part of tidying up the codebase.

I don't think the auxiliary changes for segwit are great: adding a special case `-segwitheight` arg to replace the generic `-vbparams` and dropping `get_bip9_status` entirely both seem like backwards steps. If we're going to stop testing bip9 behaviour via csv or segwit activation in the functional tests, probably should add explicit tests for it using testdummy, which would then use get_bip9_status. Likewise, it seems like it would make more sense to have all soft forks always active in regtest, with exceptions for feature specific tests to ensure pre-activation behaviour is correct, rather than having that be segwit specific.

I think we should be doing [what sipa suggested in the first PR](https://github.com/bitcoin/bitcoin/pull/11398#issuecomment-335599326): "I would prefer it if a buried deployment wouldn't require all code paths that check the BIP9 status to require changing (and instead plug in the burying into the BIP9 status logic, so that it just always returns DEFINED before the bury height, and always ACTIVE after?)."

jnewbery commented:
Thanks for the review @ajtowns 

> If we want to change the way getblockchaininfo reports softforks (and I think we do), we should split that out as a separate PR (like #16328 proposed) and review those changes independently, rather than forcing them in as part of tidying up the codebase.

This PR changes the way that softforks are reported, so that _buried_ softforks are reported as such. As you point out, this is a breaking change to the `getblockstats` API, but one that I think makes the reporting clearer (why does `softforks[0]['reject']['status'] = false` currently mean that bip34 is active?!) The new format is also extensible and allows alternative activation methods in future, if those are required.

That API change is good by itself, but doesn't make much sense outside the context of burying the bip9 softforks. The later commits in this PR change the 'activation method' of those softforks from bip9 to buried, so is another change to how the softforks are reported in `getblockchaininfo`. I think it makes sense to look at the changes as a whole. That's why these commits are contained in the same PR.

#16328 was originally opened as just the first commit from PR, but Marco then made significant changes to the code, meaning that this, #16328 and #16229 were three competing and incompatible PRs to implement overlapping changes. I think it makes sense to focus review on just one PR rather than spread it across many PRs (which probably results in none of them getting merged).

> It includes substantial user-visible changes, "breaking" the getblockchaininfo API

Both #16328 and #16229 also break `getblockchaininfo` API. As explained above, I think it makes sense to break it once to clean up that interface and make it extensible for future soft forks.

> It drops chunks of functional test coverage, in particular, after this PR, I don't think we have functional tests checking BIP9 activation works anymore

Yes! That's part of the point. We have test code that is entirely redundant and useless now that all bip9 forks are activated. If schnorr/taproot or future softforks use bip9, we can easily add tests for them.

> ... evidenced by the number of rebases this PR has required, compared to the lack of rebases the alternative PR #16229 has needed. 

This PR has required many rebases because it (and its predecessors) have been open for almost two years. I haven't checked how frequently #16229 has needed rebase, but a lot of the touched code seems comparable (RPC, validation). This PR contains quite a few changes to the functional tests which has contributed to the number of rebases.

> I think we should be doing what sipa suggested in the first PR: "I would prefer it if a buried deployment wouldn't require all code paths that check the BIP9 status to require changing (and instead plug in the burying into the BIP9 status logic, so that it just always returns DEFINED before the bury height, and always ACTIVE after?)."

Defining (and burying) soft forks should be a very infrequent activity. After this PR, the code changes required to bury a softfork should be minimal.

I'll take a look at the inline code comments tomorrow.

TheBlueMatt reviewed (commented):
Concept ACK, though I have zero opinion on this vs #16229. If we want something fancier, lets do it, otherwise this is fine.

laanwj commented on src/chainparams.cpp:
```diff
@@ -359,11 +339,22 @@ class CRegTestParams : public CChainParams {
+    if (gArgs.IsArgSet("-segwitheight")) {
+        int64_t height = gArgs.GetArg("-segwitheight", consensus.SegwitHeight);
+        if (height < -1 || height >= std::numeric_limits<int>::max()) {
+            throw std::runtime_error(strprintf("Activation height %ld for segwit is out of valid range. Use -1 to disable segwit.", height));
```
As I understand this is not called during validation, but during argument parsing at start. This follows the example of the other parsing failures later in this function.

jnewbery replied:
Correct, this isn't called during validation, only during initialization:

```
→ bitcoind -segwitheight=99999999999999999999999999
Error: Activation height 9223372036854775807 for segwit is out of valid range. Use -1 to disable segwit.
```

ajtowns replied:
The throw that occurs via validation.cpp is the `std::logic_error` when calling `VersionBits(Tip)ActivationHeight()` on a deployment that isn't locked in or active

jnewbery replied:
I've now removed that `std::logic_error`

jnewbery commented on src/validation.cpp:
```diff
@@ -4626,6 +4626,22 @@ int VersionBitsTipStateSinceHeight(const Consensus::Params& params, Consensus::D
     return VersionBitsStateSinceHeight(::ChainActive().Tip(), params, pos, versionbitscache);
 }
 
+int64_t VersionBitsActivationHeight(const Consensus::Params& params, Consensus::DeploymentPos pos)
```
done

ajtowns replied:
Making Period() a public member is cheating :)

jnewbery replied:
urgh. _None_ of the BIP9 code is currently used. I've now moved logic from `rpc/blockchain.cpp` into `validation.cpp` after @sdaftuar's comment here: https://github.com/bitcoin/bitcoin/pull/16060#discussion_r305502345, then pulled it into `versionbits.cpp` because of your comment here https://github.com/bitcoin/bitcoin/pull/16060#discussion_r308028983.

I can add even more machinery to versionbits.cpp to avoid making `Period()` public, but I don't understand where we're going with this. Once again, this is code that is _not_ expected to be run. This PR was supposed to be reducing the code complexity, but successive comments have asked for more and more code to be added to cover unexercised code paths.

ajtowns replied:
Sorry, that was just meant as a tease, not a criticism or a suggestion to change it. The ThresholdConditionChecker API isn't great.

jnewbery commented on src/validation.cpp:
```diff
@@ -4626,6 +4626,22 @@ int VersionBitsTipStateSinceHeight(const Consensus::Params& params, Consensus::D
+        case ThresholdState::ACTIVE:
+            return since_height;
+        default:
+            throw std::logic_error("VersionBitsActivationHeight called for non-ACTIVE non-LOCKED-IN deployment");
```
done

ajtowns replied:
versionbits.cpp::VersionBitsActivationHeight() still has the throw rather than returning -1 as of 41fd9d327f544d0e2f1f419a9ca70282ef5d30d2 though there's a test to see if it returns -1 in rpc/blockchain.cpp

jnewbery replied:
fixed

jnewbery commented on src/validation.cpp:
```diff
@@ -1632,7 +1632,8 @@ static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consens
 
-    if (IsNullDummyEnabled(pindex->pprev, consensusparams)) {
+    // Start enforcing BIP147 NULLDUMMY (activated simultaneously with segwit)
+    if (pindex->nHeight >= consensusparams.SegwitHeight) {
```
done

jnewbery commented:
Pushed a commit to address @ajtowns feedback. I'll squash it into the appropriate commits if people are happy with it.

jnewbery commented:
Added another commit that removes the `height` return value entirely from BIP9 forks which are locked_in. That avoids having to add a bunch of BIP9 machinery for very little benefit.

jnewbery commented:
I think all review comments are addressed. @sdaftuar @TheBlueMatt : can you re-review with the fixup commits? If you're happy, I'll squash.

ajtowns commented on src/init.cpp:
```diff
@@ -516,6 +516,7 @@ void SetupServerArgs()
     gArgs.AddArg("-maxtipage=<n>", strprintf("Maximum tip age in seconds to consider node in initial block download (default: %u)", DEFAULT_MAX_TIP_AGE), true, OptionsCategory::DEBUG_TEST);
     gArgs.AddArg("-printpriority", strprintf("Log transaction fee per kB when mining blocks (default: %u)", DEFAULT_PRINTPRIORITY), true, OptionsCategory::DEBUG_TEST);
     gArgs.AddArg("-printtoconsole", "Send trace/debug info to console (default: 1 when no -daemon. To disable logging to file, set -nodebuglogfile)", false, OptionsCategory::DEBUG_TEST);
+    gArgs.AddArg("-segwitheight=<n>", "Set the activation height of segwit. -1 to disable. (regtest-only)", false, OptionsCategory::DEBUG_TEST);
```
This probably should be in chainparamsbase.cpp? At least that's where the help for -vbparams is.

jnewbery replied:
moved

jnewbery commented:
Squashed the fixup commits and addressed @ajtowns comment here: https://github.com/bitcoin/bitcoin/pull/16060#discussion_r309561920

Pre-squashed branch is here: https://github.com/jnewbery/bitcoin/tree/pr16060.1

This is ready for re-review

jnewbery commented:
rebased

maflcko reviewed (commented):
partial ACK aece4b20841f30afa1d4a988034d07fd660b19df

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK aece4b20841f30afa1d4a988034d07fd660b19df
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUj3XAwAutgRJEbg4jm5hZzCo6MigswK3XZbtbOlEdZWCMTDKtcopN+VU6fshU8j
lWEzufNDXu/LYQyFfp2ljZPKPkLeDg+/WmLvyGsoaygMSxb8JHYbJUv9nwiLNkBG
J1kRtgfMUombAL/TyZEeeDyD8i7CRFAy7ebi7k3HRntISqJ4gF/2qcUT3PHGjV9f
89srx9ju9d/f6CAh7V8WHBJWqhEHWNzxsHg9PrG/kmj5h6JHWjpgJFArdBZYiznF
XJAnR/MPePt9O8JTTcaxlkG8uuDuNyOYiQj+SQjc11e1ZRiBatBugZ0v36Xrk7qy
wQQHFP0uWz7RbJFao1TPiRkJC2PUiDVVj1xJHkZTBBxS840AAFZ8QjiY2Zb06yIb
2GQK25KGiF4RVkSxIaXS/Knaq7iVZQev3wwYwcCEKkD4/9r4JQYsmQBTHN5epmsW
ov2cF37OAsCswjorn2bWJfjmZ5DzBeSi52RVIfbl3EQogJvEfFUXb/J7Z+X9olkU
qA5NhpYa
=qhoe
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `158b950bfd063584b8e84c3c60eb8c5aedcfdf52005b2c7784a8cb0cef5d418b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108158b950bfd063584b8e84c3c60eb8c5aedcfdf52005b2c7784a8cb0cef5d418bf010cbc16b9d751dd275da2d87c0800367fc08fff0106b3392a027a0f8d9be58223d798251e808f12087a271dbd6a58fafd43804c71b542543cd14660c6a9ebc5dc82a24e22518227f08f1045d44888af0088146c3e08f4ead810083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01085a0c4c3e5e6228ad0c94bb69d8b848608f1045d44888af0083f704d911c818d880083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0108f7f122c1a9f1ff5beb43fe419f4c29e08f120a4a73bc9d01e6906e45d79489dfcf1d02e82e0518bb7eaef532a7d32226cb4b508f1045d44888af00887fe42248c50059c0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010984c8544947bc9f02da1fec14a380b0f08f1201d315e8a8f5b956bcb230cc5c0f757e4cb393abc458c87ea437043f0b97367dc08f1045d44888af0085f746023f918adea0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

ajtowns commented on src/init.cpp:
```diff
@@ -1670,12 +1670,9 @@ bool AppInitMain(InitInterfaces& interfaces)
+    if (chainparams.GetConsensus().SegwitHeight != std::numeric_limits<int>::max()) {
+        // Advertise witness capabilities.
+        // The option to not set NODE_WITNESS is only used in the tests and should be removed.
         nLocalServices = ServiceFlags(nLocalServices | NODE_WITNESS);
```
I don't think this works -- p2p_segwit launches some nodes with `-segwitheight=120` and one with segwit disabled, and expects them to follow the same blockchain. But if you set NODE_WITNESS then when the first node thinks segwit is enabled at block 120, it'll send the witness data to the other node, which will then give an error message, "AcceptBlock FAILED (unexpected-witness, ..." and fail the test.

The test case is essentially testing that upgrades work, though, which would be more useful if the non-segwit node was actually running old code, rather than just simulating it... I think the only changes needed to p2p_segwit for this are:

    * drop `test_upgrade_after_activation()`
    * in `test_getblocktemplate_before_lockin()` set `node=self.nodes[0]`, drop the for loop, and drop the dead code for `node == self.nodes[2]`.

I don't think this needs to be done in this PR though.

ajtowns commented on src/validation.h:
```diff
@@ -385,9 +385,6 @@ bool TestBlockValidity(CValidationState& state, const CChainParams& chainparams,
 /** Check whether witness commitments are required for block. */
 bool IsWitnessEnabled(const CBlockIndex* pindexPrev, const Consensus::Params& params);
 
-/** Check whether NULLDUMMY (BIP 147) has activated. */
```
*sigh* Probably should update the comment for `IsWitnessEnabled` above...

Might be good to note that segwit is enforced for all transactions, and this function is only used for nulldummy and block-level things (ie, working out if you've upgraded and need to reprocess segwit blocks that you didn't download witness data for, if you should generate witness data for getblocktemplate, and so you can accept old blocks from non-witness peers)

jnewbery replied:
comment updated

jnewbery commented on test/functional/feature_csv_activation.py:
```diff
@@ -267,7 +236,7 @@ def run_test(self):
         self.send_blocks(test_blocks)
 
         self.log.info("Not yet advanced to ACTIVE, height = 574 (will activate for block 576, not 575)")
```
done, and fixed a bunch of other comments in the test.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1183,54 +1183,48 @@ static UniValue verifychain(const JSONRPCRequest& request)
+    // For buried deployments.
+    // A buried deployment is one where the height of the activation has been hardcoded into
+    // the client implementation long after the consensus change has activated. See BIP 90.
+    // Buried deployments with activation height value of std::numeric_limits<int>::max() are hidden.
```
I've taken your suggestion. Thanks

jnewbery commented:
Updated to address @MarcoFalke and @ajtowns comments. Thanks for the review!

ariard reviewed (commented):
Reviewed c64c9b5

Mostly grep'ed `versionbits` refs, which should be updated IMO.

Tested getblockchaininfo:
* tip on testnet
* tip on testnet, with a segwit activation height far in the future
* at 7e33aea, to see if bip 9 object display well
* csv activation at height 431 on regtest

With this PR, bip 9 logic isn't tested anymore, is there an easy way to write a dummy test for it without polluting code paths ? Don't think so..

(also add comments for versionbits.h in https://github.com/ariard/bitcoin/commit/fca79982227978c4b5ecbb7489754e54fb179696, obviously no one wants to weigh this already wide-scoped PR but it may be relevant on its own)

ajtowns commented on test/functional/feature_cltv.py:
```diff
@@ -69,14 +69,11 @@ def skip_test_if_missing_module(self):
     def test_cltv_info(self, *, is_active):
-        assert_equal(
-            next(s for s in self.nodes[0].getblockchaininfo()['softforks'] if s['id'] == 'bip65'),
+        assert_equal(self.nodes[0].getblockchaininfo()['softforks']['bip66'],
```
Good catch; good example of why not to use 'bipXXX' as identifiers imo. Also the first check should be something like:

```diff
-        self.test_cltv_info(is_active=True)
+        self.test_cltv_info(is_active=False) # Not active as of current tip and next block does not need to obey rules
```

jnewbery replied:
Great catch. This was careless on my part (copy-pasting from feature_dersig.py to feature_cltv.py).

I've fixed this and added the comment as suggested by AJ.

ajtowns commented on src/validation.cpp:
```diff
@@ -3045,14 +3046,8 @@ bool CheckBlock(const CBlock& block, CValidationState& state, const Consensus::P
-{
-    LOCK(cs_main);
-    return (VersionBitsState(pindexPrev, params, Consensus::DEPLOYMENT_SEGWIT, versionbitscache) == ThresholdState::ACTIVE);
+    int height = pindexPrev == nullptr ? 0 : pindexPrev->nHeight + 1;
```
I don't think that works for `UpdateUncommittedBlockStructures` or `BlockAssembler::CreateNewBlock`, so you'd still need this function to be available.

ariard replied:
Right these functions test for the next block to be generated/submitted based on the parent,  for them you may pass `pindexPrev->nHeight + 1` to `IsWitnessEnabled`, should work but need to refactor also callers ?

jnewbery replied:
Right, I think this could also be refactored to take a `height` parameter and just return whether segwit should be enforced for a block at that height.

We can leave that for a follow-up PR if people think it's worthwhile.

ajtowns replied:
If it was going to take a height parameter, might as well just have the test in place directly. I don't think it's worth fiddling with for just segwit.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1240,18 +1235,18 @@ static UniValue BIP9SoftForkDesc(const Consensus::Params& consensusParams, Conse
+    UniValue rv(UniValue::VOBJ);
+    rv.pushKV("type", "bip9");
+    rv.pushKV("bip9", bip9);
+    if (ThresholdState::ACTIVE == thresholdState) {
```
This was in the original version of the PR, but Suhas didn't like that it leaked details of the BIP9 implementation into the RPC code (https://github.com/bitcoin/bitcoin/pull/16060#discussion_r305502345). I tried adding additional BIP9 code to expose just the activation height (https://github.com/bitcoin/bitcoin/pull/16060#issuecomment-516898719), but it didn't seem worth it so I ended up removing that (https://github.com/bitcoin/bitcoin/pull/16060#discussion_r308028983).

laanwj replied:
imo: let's try not to circle around too much here, and get the consensus burying part merged, the RPC part can be improved in later PRs

jnewbery commented on src/validation.cpp:
```diff
@@ -3187,7 +3187,7 @@ static bool ContextualCheckBlock(const CBlock& block, CValidationState& state, c
 
     // Start enforcing BIP113 (Median Time Past) using versionbits logic.
```
fixed

jnewbery commented on test/functional/feature_bip68_sequence.py:
```diff
@@ -341,7 +341,7 @@ def test_nonzero_locks(orig_tx, node, relayfee, use_height_lock):
     # being run, then it's possible the test has activated the soft fork, and
     # this test should be moved to run earlier, or deleted.
     def test_bip68_not_consensus(self):
```
fixed

jnewbery commented on test/functional/feature_csv_activation.py:
```diff
@@ -187,45 +188,14 @@ def run_test(self):
-        self.log.info("Test that the csv softfork is DEFINED")
-        assert_equal(get_bip9_status(self.nodes[0], 'csv')['status'], 'defined')
-        test_blocks = self.generate_blocks(61, 4)
+        # Activation height is hardcoded
```
fixed

jnewbery commented on test/functional/feature_dersig.py:
```diff
@@ -52,14 +52,11 @@ def skip_test_if_missing_module(self):
     def test_dersig_info(self, *, is_active):
-        assert_equal(
-            next(s for s in self.nodes[0].getblockchaininfo()['softforks'] if s['id'] == 'bip66'),
+        assert_equal(self.nodes[0].getblockchaininfo()['softforks']['bip66'],
```
this also tests the height and type, so I'll leave it as it is.

jnewbery commented on src/chainparams.cpp:
```diff
@@ -275,6 +267,7 @@ class CRegTestParams : public CChainParams {
         consensus.BIP34Hash = uint256();
         consensus.BIP65Height = 1351; // BIP65 activated on regtest (Used in functional tests)
         consensus.BIP66Height = 1251; // BIP66 activated on regtest (Used in functional tests)
+        consensus.CSVHeight = 432; // CSV activated on regtest (Used in rpc activation tests)
```
hmmm, it's an interesting point, but I don't think there's an issue here. Previously, feature_csv_activation.py activated CSV at height 576, so that was still before bip 65/66 activation.

jnewbery commented:
Thanks for the review @ariard . I think I've addressed all your comments.

ajtowns commented:
ACK 20e60006de5015e057121b74b623698e96c0bfae ; code review, compiles, tests pass, reports same activation heights as master for segwit/csv on testnet/mainnet. I've not thoroughly reviewed the changes in the python tests.

Nits:
 * the first commit ("tidy up reporting...") only touches validation.h and validation.cpp in order to add/remove blank lines
 * the constant 9223372036854775807 in the tests would probably be clearer as `0x7fffffffffffffff`(or even 2**63-1)

I think release notes are warranted for the change to getblockchaininfo and probably getblocktemplate. I'm not sure the change to getblocktemplate (no longer returning "segwit" in the "rules" field of the result) are okay, but Luke didn't complain, so hopefully it's fine. Both those seem fine to worry about later.

I tried drafting up a bip9 functional test making use of testdummy on regtest, but I'm not convinced it makes sense to do that directly rather than improving our bip9 implementation or improving bip9 first. Either way, that can be left to a later PR.

jnewbery committed:
[docs] Add release notes for burying bip 9 soft fork deployments

jnewbery commented:
Thanks @ajtowns . I've force-pushed a fix to those nits and also release notes.

I also intend to send a post to the mailing list if this gets merged (I don't think a full BIP is necessary).

> I tried drafting up a bip9 functional test making use of testdummy on regtest, but I'm not convinced it makes sense to do that directly rather than improving our bip9 implementation or improving bip9 first. Either way, that can be left to a later PR.

Very happy to review any tests or changes to BIP 9 you PR.

ajtowns commented:
>  (I don't think a full BIP is necessary)

I agree; IMO that the heights are already included in https://github.com/bitcoin/bips/blob/master/bip-0009/assignments.mediawiki means the BIP side of things is already covered.

ariard commented on test/functional/feature_csv_activation.py:
```diff
@@ -340,10 +304,10 @@ def run_test(self):
 
-        # 1 more version 4 block to get us to height 575 so the fork should now be active for the next block
-        test_blocks = self.generate_blocks(1, 4)
+        # 1 more version 4 block to get us to height 432 so the fork should now be active for the next block
```
nit : maybe you could add an assert not active before activation, just to be sure everything is clean after the send_blocks/invalidateblock sequence

jnewbery replied:
done

ariard reviewed (approved):
ACK dbe75a0, reviewed code, tests ok. I've had a look on python tests, comments have been addressed since last time and generally they look good.

jnewbery commented on src/validation.cpp:
```diff
@@ -1644,8 +1644,8 @@ static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consens
 
-    // Start enforcing BIP68 (sequence locks) and BIP112 (CHECKSEQUENCEVERIFY) using versionbits logic.
-    if (VersionBitsState(pindex->pprev, consensusparams, Consensus::DEPLOYMENT_CSV, versionbitscache) == ThresholdState::ACTIVE) {
+    // Start enforcing BIP68 (sequence locks) and BIP112 (CHECKSEQUENCEVERIFY)
```
good spot. Fixed!

jnewbery commented:
Thanks for the review @ariard - I've addressed your comments.

ajtowns commented:
ACK e78aaf41f43d0e2ad78fa6d8dad61032c8ef73d0 ; checked diff to previous acked commit, checked tests still work

ariard commented:
ACK e78aaf4, check diff, run the tests again and successfully activated csv/segwit heights on mainnet as expected.

MarcoFalke commented:
ACK e78aaf41f43d0e2ad78fa6d8dad61032c8ef73d0 (still didn't check if the mainnet block heights are correct, but the code looks good now)

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK e78aaf41f43d0e2ad78fa6d8dad61032c8ef73d0 (still didn't check if the mainnet block heights are correct, but the code looks good now)
-----BEGIN PGP SIGNATURE-----

iQGyBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhaqwv42/MfP5paFChypsDXr5M3MZtfZyD36YtC9JCjdtdBff2zKLKONh1PvrYj
d7MuAJ6fgsaS7qE4GIbCo77EDzJziqTsgoruoWckqzLrHgr2Bt4gaxEpciQhUrVD
sR4GjnWuVPkawnIPkr2+hJkScU2/fFGVnVzF1Hg0A7mIzkr9xdt1ofJvb9GQ4O2Y
fteE78HpUAgwvmoudddIsRDKGui/0tTClzhx8D+1cVz7MtSU7+2bmpcvdhn++vHl
q6apUyGTzOFZtC4cJeV/1cv7Jka3M59IC/U05zZU6CD/ocDM3xrXY3z+tZCJEUIJ
MdwWcqBuaK01ytvEr9ZmQ4cutLZFIkVPkz7FgayVU7XZ1X9666TGA1vnY/k1oh7e
vl5+QZl+2bgleQRLot1vGRM5WXzHyksQUGT/XFTmumYmBY36zu4+pt4cUtLabnhW
G3S2joUyMM9JonKCgYGcyPx3XUvLan5E5pZp092RXGoRB2f2T3ZwwDqwhIfmDFhj
1HTWBpo=
=HDUJ
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `f62d1be3493fe779f9feb2f59b46db1f25e6e089a7d58db2e083156c727bb44a  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108f62d1be3493fe779f9feb2f59b46db1f25e6e089a7d58db2e083156c727bb44af0102204b546ea5e8758c7769b6091687a8b08fff01010c96dedcbaa4ed2c5a67f2723822c1c08f1045d55b9e9f00898c04f1a19162f090083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01013829c7b2a2d8f2684c46dc40239fded08f1045d55b9eaf0088fab400a33d879ef0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010c4291d96ddf515147d7e316c8591c63508f1045d55b9e9f00888e990f8ac74d8ef0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f478d73358a27d0922f125d62c95713a08f1045d55b9eaf008d01d4fe967c1629d0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>
